### ISSUE TITLE
Mergers get stuck when running in parallel

### ISSUE THREAD

**Issue by manuelma** [Spine-project/Spine-Toolbox#1769]
When multiple mergers want to write to the same output db, sometimes they get stuck. I believe the reason is writing to SQLite dbs from different processes is risky, although I couldn't find anything about it online. It's just a hunch.

One solution could be to make them write on the main process via Spine DB server. Will try to implement that.

**Comment by jkiviluo** [comment]
Would using the 'main process' lock the Toolbox interface? Maybe I misunderstand, but would one option be to start one new thread to handle all the imports? On the other hand, if the importer does lot of work before writing - could that work be in parallel threads, but the writing in one 'writer thread' that lines up the writes to be executed in series?

**Comment by manuelma** [comment]
You got it @jkiviluo that's exactly what we need to do: have a writer process that's not the UI process and does all the writing. I think I'm close to achieve that.

**Comment by manuelma** [comment]
I managed to accomplish the above after some struggling with Python multiprocessing - I'm now satisfied with the result so will close the issue, please reopen if needed.

To summarize, changes are:
- Both Importer and Merger use the Spine DB Server to write the 'import data' to the DB.
- All Spine DB servers run in a dedicated Process. This ensures that no two processes attempt to write to the same DB simultaneously, which in turns prevents any deadlock.

One final thought: since all writings are now controlled by the Spine DB Server, it shouldn't be too hard to implement an extra layer to control write order (another issue we had pending from the Spine Project).

**Comment by manuelma** [comment]
This is still doesn't work if we use other process than the main process to start servers. I just changed it so we use the UI process for servers which prevents all deadlocks but does block the UI a little bit (although not dramatically) when merger execute. Will try and figure out what's going on.

**Comment by manuelma** [comment]
It seems that the issue was related to starting the servers' process on demand to support windows (see https://github.com/Spine-project/Spine-Toolbox/issues/1771)

As result, the process could be started from any thread which defeated the purpose. Now it seems to work better. Will close and reopen if I find more problems.
